crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -2187,17 +2207,40 @@ pub trait CheckpointServiceNotify {
     fn notify_checkpoint(&self) -> SuiResult;
 }
 
-/// This is a service used to communicate with other pieces of sui(for ex. authority)
+enum CheckpointServiceState {
+    Unstarted((CheckpointBuilder, CheckpointAggregator)),
+    Started,
+}
+
+impl CheckpointServiceState {
+    fn take(&mut self) -> (CheckpointBuilder, CheckpointAggregator) {
+        let mut state = CheckpointServiceState::Started;
+        std::mem::swap(self, &mut state);
+
+        match state {
+            CheckpointServiceState::Unstarted((builder, aggregator)) => (builder, aggregator),
+            CheckpointServiceState::Started => panic!("CheckpointServiceState is already started"),
+        }
+    }
+}
+
 pub struct CheckpointService {
     tables: Arc<CheckpointStore>,
     notify_builder: Arc<Notify>,
     notify_aggregator: Arc<Notify>,
     last_signature_index: Mutex<u64>,
+    // The highest sequence number that had already been built at the time CheckpointService
+    // was constructed
+    last_built_seq: CheckpointSequenceNumber,
+    // A notification for each time a new checkpoint is built.
```

Comment:
this comment is confusing me because you are using a `watch`, which I don't believe is guaranteed to report every sent value to a receiver if the receiver is too slow? (but "each time" makes it seem like it is important for every value to be received?) or am I wrong about how watch works?

Reply:
Good point. Reworded the comment.